### DISCUSSION TITLE
Adapting the forward simulation of the crosshole tomography example to 3D

### DISCUSSION THREAD

**Discussion by sminize** [gimli-org/pyGIMLi#400] [Q&A]
Hello everyone,
I want to adopt the forward simulation of the cross-hole tomography example to 3D. (https://www.pygimli.org/_examples_auto/seismics/plot_3_crosshole_tomography.html)
I tried the following: 

**First approach; adapt the mesh generation to 3D:**

```
bh_spacing = 10.5


world = mt.createCube(start=np.array([0, 0,0]), end=np.array([bh_spacing, bh_spacing, bh_spacing]),
                           marker=0)

c0 = mt.createCylinder(pos=(4.5, 4.5), radius= 1.75,height=10.5, nSegments=25, marker=1)
c1 = mt.createCylinder(pos=(6.5, 8.0), radius=0.75,height=10.5, nSegments=25, marker=2)
geom = world + c0 + c1
for sen in sensors:
    geom.createNode(sen)

mesh_fwd = mt.createMesh(geom, quality=34, area=0.25)

```
I get the following error message:

```
Opening /tmp/tmpvb9tbni_.poly.
The input surface mesh contains self-intersections. Program stopped.

```

**My second approach was to extrude the 2D mesh like this:**

```
z = np.linspace(0,10.5,5)
mesh_fwd = pg.meshtools.extrudeMesh(mesh_fwd, a=z)
```
When I start the forward simulation I get:
`pyGIMLi - WARNING - cell type unknown`

How can I fix this?

Thank you for your help.

All the best,

Simon Schmid

**Comment by carsten-forty2** [comment]
Hi Simon,

the first fails because `world`, `c0` and `c1` are intersecting each other. Curently only non.intersecting objectes can be summed. I recommand to take a look at the geometry in paraview (until we fixed the buildin 3d viewer) `geom.exportVTK('p1')`

![1](https://user-images.githubusercontent.com/6043771/175134282-9e33ea42-c909-4a8e-9743-0503a0169f0d.png)

The second works as expected here .. I just needed to copy the original 2D `mesh_fwd` from the mentioned example:
![2](https://user-images.githubusercontent.com/6043771/175135312-577455c3-2ca9-4e79-9517-e9bf4c24222a.png)

Please ensure you have the most recent version of `pygimli` installed, if you have .. please give the output of pg.version()

Cheers

**Reply by sminize** [reply]
Hi Carsten, 
thank you very much for your help. 
I got pygimli Version 1.2.6 installed. 
Am I correct that, since intersection objects are not allowed in 3D, creating a cubic mesh with cylindrical inserts is not possible in 3D by using the sum of mesh objects?



The mesh creation was for me also possible by extrusion. But the forward Simulation gave me a warning, which was:

`pyGIMLi - WARNING - cell type unknown`

It somehow does not like the shapes of the cells. Is there a way to fix this?

By the way, how can I do the 3D visualization of the mesh?

Thank you for your help,

All the best,

Simon

**Reply by carsten-forty2** [reply]
While intersection of objects are not yet possible with meshtools, objects can touch each other. It should be possible to create a valid 3D PLC if you align the cylinder inside the cube. With intersection I mean the intersection of object surfaces.

Sorry, I did not read the warning came from forward calculation. Its possible that traveltime calculations are not yet implemented for prisms. I need to check this. Lets see if the first way works for you, I think tetrahedrons may work.

For 3D Visualisation you can try `pg.show()` since there are very recent changes in the dev branch (my 2nd figure). Or you need to wait for the next release and export the PLC or the mesh in `vtk` file format and open these files in Paraview https://www.paraview.org/ (my 1st Figure)

**Reply by sminize** [reply]
Hi Carsten,
with that approach, it worked. I just had to adapt the `createCylinder` function a bit in order to align the cylinders in the cube.

Thanks a lot

All the best,

Simon

**Comment by sminize** [comment]
Here is what I did so far using the extrusion method:

```
import matplotlib.pyplot as plt
import numpy as np

import pygimli as pg
import pygimli.meshtools as mt
from pygimli.physics.traveltime import TravelTimeManager, createCrossholeData


pg.utils.units.quants['vel']['cMap'] = 'inferno_r'
################################################################################
# Next, we build the crosshole acquisition geometry with two shallow boreholes.

# Acquisition parameters
bh_spacing = 10.5
bh_length = 10.5
sensor_spacing = 1.5

world = mt.createRectangle(start=np.array([0, 0,]), end=np.array([bh_spacing, bh_spacing]),
                           marker=0)
sensors=np.array([[0, 1.635,5], [0, 5.25,5], [0, 8.865,5],[1.635,0,5], [5.25,0,5], [8.865,0,5]])
receivers=np.zeros((21,3))
receivers1=np.zeros((21,3))
receivers[:,0]=10.5
receivers1[:,1]=10.5
receivers1[:,2]=5
dx=0.425
for i in range(21):
    receivers[i,1]=1.0 + i * dx
    receivers1[i,0]=1.0 + i * dx
receivers=np.concatenate((receivers,receivers1),axis=0)
sensors = np.concatenate((sensors,receivers),axis=0)

c0 = mt.createCircle(pos=(4.5, 4.5), radius= 1.75, nSegments=25, marker=1)
c1 = mt.createCircle(pos=(6.5, 8.0), radius=0.75, nSegments=25, marker=2)

geom = world + c0 + c1
for sen in sensors:
    geom.createNode(sen)

mesh_fwd = mt.createMesh(geom, quality=24, area=0.55)
model = np.array([6377.08, 2690.0, 2690.0])[mesh_fwd.cellMarkers()]
pg.show(mesh_fwd, model,
        label=pg.unit('vel'), cMap=pg.cmap('vel'), nLevs=3, logScale=False)

z = np.linspace(0,10.5,25)
mesh_fwd = pg.meshtools.extrudeMesh(mesh_fwd, a=z)

from itertools import product
rays = list(product(np.array(range(3)), 6+np.array(range(21))))
rays2 = list(product(3+np.array(range(3)), 6+21+np.array(range(21))))

rays=rays+rays2

scheme = pg.DataContainer()

# Add sensors
for sen in sensors:
    scheme.createSensor(sen)

# Add measurements
rays = np.array(rays)
scheme.resize(len(rays))
scheme.add("s", rays[:, 0])
scheme.add("g", rays[:, 1])
scheme.add("valid", np.ones(len(rays)))
scheme.registerSensorIndex("s")
scheme.registerSensorIndex("g")

tt = TravelTimeManager()
data = tt.simulate(mesh=mesh_fwd, scheme=scheme, slowness=1./model,
                   secNodes=4, noiseLevel=0.001, noiseAbs=1e-5, seed=1337)
```


The simulation, however, does not work with this mesh. Do you have an idea how to fix this?

All the best,

Simon
